homework-submission/main.js

```diff
@@ -0,0 +1,147 @@
+'use strict';
+
+const todoArr = [
+  {
+    id: 1,
+    description: 'Get out of bed',
+    deadline: '2017-09-11',
+    done: true,
+  },
+  {
+    id: 2,
+    description: 'Brush teeth',
+    deadline: '2017-09-10',
+    done: false,
+  },
+  {
+    id: 3,
+    description: 'Eat breakfast',
+    deadline: '2017-09-09',
+    done: false,
+  },
+];
+
+const Title = ({ text }) => <h1 className="title">{text}</h1>;
+
+const Input = ({ type, name }) => {
+  return (
+    <div className="input-section">
+      <label className="label" htmlFor={name}>
+        {name}:
+      </label>
+      <input className="input" type={type} name={name} required={type !== 'checkbox'} />
+    </div>
+  );
+};
+
+const Button = ({ type, text, clickHandler }) => {
+  return (
+    <button className="button" type={type} onClick={clickHandler}>
+      {text}
+    </button>
+  );
+};
+
+const ListItem = ({ clickHandler, todo }) => {
+  const date = new Date(todo.deadline).toLocaleDateString('en-EN', {
+    weekday: 'short',
+    month: 'short',
+    day: 'numeric',
+    year: 'numeric',
+  });
+  return (
+    <li onClick={clickHandler} id={todo.id} className={`list-item ${todo.done}`}>
+      {todo.description}, {date}
+    </li>
+  );
+};
+
+const AddNewTodo = ({ submitHandler }) => {
+  return (
+    <section className="form container">
+      <Title text="Add a new todo" />
+      <form className="form" onSubmit={submitHandler}>
+        <Input type="text" name="description" />
+        <Input type="date" name="deadline" />
+        <Input type="checkbox" name="done" />
+        <Button type="submit" text="Add New Todo" />
+      </form>
+    </section>
+  );
+};
+
+const TodoList = ({ todoArr, liClickHandler, buttonClickHandler, clickBehaviour }) => {
+  return (
+    <section className="list container">
+      <Title text="Todo List" />
+      <Button clickHandler={buttonClickHandler} text={clickBehaviour} />
+      <ul className="list">
+        {todoArr.map(todo => (
+          <ListItem key={todo.id} clickHandler={liClickHandler} todo={todo} />
+        ))}
+      </ul>
+    </section>
+  );
+};
+
+class Main extends React.Component {
+  state = { todoArr: [...todoArr], behaviour: 'mark' };
+
+  addNewTodo(event) {
+    event.preventDefault();
+    const currentTodos = [...this.state.todoArr];
+    const newTodo = {
+      id: currentTodos[currentTodos.length - 1] ? currentTodos[currentTodos.length - 1].id + 1 : 1,
+      description: event.target.description.value,
+      deadline: event.target.deadline.value,
+      done: event.target.done.checked,
+    };
+    currentTodos.push(newTodo);
+    this.setState({ todoArr: [...currentTodos] });
```

Comment:
You are already using the spread operator (`...`) and the new array operator (`[]`) to create a copy of the state. There is no need to do this again here on line 100 :-) This just results in unnecessary operations/performance/memory issues.